### DISCUSSION TITLE
How to use the library in a client-server application: how to correctly embed telegram user authentication?

### DISCUSSION THREAD

**Discussion by Kalinin-Andrey** [xelaj/mtproto#127] [Q&A]
**Как использовать библиотеку в клиент-серверном приложении: как корректно встроить аутентификацию телеграм-пользователей?**

Всем привет!

**Во-первых, спасибо за труд авторам библиотеки!**

Мне пока не понятно как вашу библиотеку использовать в моём клиент-серверном приложении.

1. Клиент (`telegram.NewClient()`) сохраняет своё состояние в сессионом файле и соответствует чему: телеграм-приложению (со своим `AppId` и `AppHash`) или телеграм-пользователю? На сколько я понял, сессионный файл создасться один на одно телеграм-приложение, это верно?. Но является-ли этот механизм потокобезопасным? Как приложение будет работать если с ним параллельно будут логиниться и работать несколько телеграм пользователей?

2. Должен-ли клиент (`telegram.NewClient()`) создаваться при каждом запросе или достаточно его создать только один раз при иннициализации сервера?

3. Как в клиент-серверном приложении обработать случай когда "SESSION_PASSWORD_NEEDED"? В вашем примере кода между строками `client.AuthSignIn()` и `client.AuthCheckPassword()` сохраняется-ли какое-то состояние в оперативной памяти? Или можно разнести этот код в своём приложении по разным обработчикам запросов АПИ?
	Например:
		- сделать роут `/auth-sign-in`, который принимает код, и в данном случае возвращает ошибку "SESSION_PASSWORD_NEEDED", при которой нужно вызывать следующий роут
		- роут `/auth-check-password`, который принимает пароль. Т.е. в этом обработчике будет только такой код:
```
	accountPassword, err := client.AccountGetPassword()
	dry.PanicIfErr(err)

	// GetInputCheckPassword is fast response object generator
	inputCheck, err := telegram.GetInputCheckPassword(password, accountPassword)
	dry.PanicIfErr(err)

	auth, err = client.AuthCheckPassword(inputCheck)
	dry.PanicIfErr(err)
```

Так будет работать корректно?

4. Есть-ли возможность сохранять сессию не в файл, а в Redis, для возможности запускать несколько серверных инстансов?

Возможно, где-то можно посмотреть код реального web приложения, где используется ваша библиотека? Чтобы посмотреть реальный пример, который бы снял все подобные вопросы.

Спасибо!

=================================================

Hi all!

**First, thanks for the work of the authors of the library!**

It is not yet clear to me how to use your library in my client-server application.

1. The client (`telegram.NewClient()`) saves its state in the session file and corresponds to what: a telegram application (with its own AppId and AppHash) or a telegram user? As far as I understand, the session file will be created one by one telegram application, is that right ?. But is this mechanism thread-safe? How will the application work if several telegram users log in and work with it in parallel?

2. Should the client (`telegram.NewClient()`) be created on each request, or is it enough to create it only once, in server initialization?

3. How to handle the case when "SESSION_PASSWORD_NEEDED" in a client-server application? In your code example, between the lines `client.AuthSignIn()` and `client.AuthCheckPassword()`, is there any state stored in RAM? Or is it possible to spread this code in your application to different API request handlers?
	For example:
		- make a `/auth-sign-in` route that accepts a code, and in this case returns a "SESSION_PASSWORD_NEEDED" error, in which you need to call the next route
		- the `/auth-check-password` route, which accepts a password. Those. in this handler there will be only such code:
```
	accountPassword, err := client.AccountGetPassword()
	dry.PanicIfErr(err)

	// GetInputCheckPassword is fast response object generator
	inputCheck, err := telegram.GetInputCheckPassword(password, accountPassword)
	dry.PanicIfErr(err)

	auth, err = client.AuthCheckPassword(inputCheck)
	dry.PanicIfErr(err)
```
So will it work correctly?

4. Is it possible to save the session not to a file, but to Redis, to be able to run multiple server instances?
Is somewhere I can see the code of a real web application where your library is used? To see a real example that would remove all such questions.

Thank you!

**Comment by quenbyako** [comment]
Привет, отвечу на русском, потом переведу на английский и сделаю отдельную страничку с документацией, вопросы хорошие)

1. Есть две сущности — сессии и клиенты. Клиент это как раз AppID + AppHash, и он позволяет телеграму собирать статистику по использованию (не владельцу приложения! ожидается что разработчик сам статистику настроит, что дикость). Сессия же привязывается к отдельному пользователю И приложению. Один пользователь в одном приложении может создать сколь угодно сессий, это не воспрещается. Каждая сессия — это ключ, которым клиент и сервер обменялись. Поэтому можно на 100 устройствах создать 100 сессий с одним AppID под одного пользователя. А можно на одном устройстве, но смысла в этом нет 😁
2. Достаточно одного раза, клиент хранит в себе данные сессии и управляет ей как необходимо. 1 `telegram.Client` — 1 инстанс (= 1 коннект к серверу телеги). В одном приложении можно запустить 10 000 клиентов, библиотека поддерживает параллельное выполнение.
3. По хорошему это прям в библиотеке надо обработать, у меня руки не дошли до этого :( Но логика ровно такая как вы описали, все верно 👍
4. Да, вы можете написать кастомный интерфейс для хранилища, если не ошибаюсь, сейчас уже есть поддержка. Сохранение в файл это просто для примера.

Прошу прощения за долгий ответ, но надеюсь поможет)
